### ISSUE TITLE
`softmax` fails to commute with `permute` / `transpose` (at least for float32)

### ISSUE THREAD

**Issue by JasonGross** [pytorch/pytorch#139701]
### 🐛 Describe the bug

It should be the case that `sometensor.softmax(dim=d)` is the same as `sometensor.permute(...).softmax(dim=0).permute(...)` if the permutations are inverses.  This is not the case, at least for float32:
```python
import torch

val = torch.tensor([[7.76308536529541015625, 7.17303562164306640625, 5.13778305053710937500]], dtype=torch.float32, device='cpu')
val1 = val.softmax(dim=1, dtype=torch.float32).permute(1, 0)
val2 = val.permute(1, 0).softmax(dim=0, dtype=torch.float32)
print(val1)
print(val2)
print((val1 - val2).abs().max())
```
```
tensor([[0.6147],
        [0.3407],
        [0.0445]])
tensor([[0.6147],
        [0.3407],
        [0.0445]])
tensor(2.9802e-08)
```
Note that, e.g., scipy, does not have this issue:
```python
import numpy as np
from scipy.special import softmax

val = np.array([[7.76308536529541015625, 7.17303562164306640625, 5.13778305053710937500]], dtype=np.float32)
val -= val.max()
val1s = softmax(val, axis=1).T
val2s = softmax(val.T, axis=0)
print(val1s)
print(val2s)
print(np.max(np.abs(val1s - val2s)))
print()
print(val1 - val1s)
print(val2 - val2s)
```
gives
```
[[0.6147349 ]
 [0.34074736]
 [0.04451779]]
[[0.6147349 ]
 [0.34074736]
 [0.04451779]]
0.0

tensor([[0.0000e+00],
        [2.9802e-08],
        [0.0000e+00]])
tensor([[0.],
        [0.],
        [0.]])
```

Discrepancy originally encountered by @ronakrm (I have not looked for what floating point values give different results under transpose)

### Versions

Collecting environment information...
PyTorch version: 2.4.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.0.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.3)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.9 (main, Jul 30 2024, 20:31:18) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-15.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] eintorch==0.2.5
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.1.1
[pip3] pytorch-lightning==2.4.0
[pip3] torch==2.4.1
[pip3] torchmetrics==1.4.2
[pip3] torchvision==0.19.1
[conda] numpy                     2.0.0           py310h52bbd9b_0    conda-forge

**Comment by JasonGross** [comment]
Here's an example that fails with both float32 and float64:
`torch.tensor([[0, -0.5, float.fromhex('-0x1.205p+1')]], dtype=torch.float64, device='cpu')`
